client/pdo/client/commands/contract.py

```diff
@@ -26,42 +26,56 @@
 import pdo.common.crypto as pcrypto
 import pdo.contract as pcontract
 from pdo.common.keys import ServiceKeys
-from pdo.common.utility import valid_service_url
 from pdo.submitter.create import create_submitter
 
 logger = logging.getLogger(__name__)
 
 __all__ = [
+    'flush_contract_cache',
     'get_contract',
     'get_contract_from_context',
     'create_contract',
     'send_to_contract',
     'do_contract',
     'load_commands',
-    ]
+]
 
 ## -----------------------------------------------------------------
 ## get_contract
 ## -----------------------------------------------------------------
-__contract_cache__ = {}
+class __hashabledict__(dict) :
+    """Hashable dictionary
 
-def get_contract(state, save_file) :
-    """Get contract object using the save_file.
+    This is a relatively simple hack to make a dictionary hashable
+    so that the cache can work. It assumes that the dictionary does
+    not change. This is sufficent for the ledger configuration.
     """
+    def __hash__(self) :
+        return hash(frozenset(self))
 
-    global __contract_cache__
+@cachetools.func.ttl_cache(maxsize=32, ttl=30)
+def __get_contract__(ledger_config, save_file, data_directory) :
+    logger.debug(f'load contract from file: {save_file}')
+    return pcontract.Contract.read_from_file(ledger_config, save_file, data_dir=data_directory)
+
+def get_contract(state, save_file) :
+    """Retrieve a contract object associated with the named save file
+    """
+    data_directory = state.get(['Contract', 'DataDirectory'])
+    ledger_config = __hashabledict__(state.get(['Ledger']))
 
-    if save_file not in __contract_cache__ :
-        try :
-            data_directory = state.get(['Contract', 'DataDirectory'])
-            ledger_config = state.get(['Ledger'])
+    # expiration from the cache really means that we want the
+    # contract re-read, so force any expired contracts out of
+    # the cache before we read it
+    __get_contract__.cache.expire()
```

Comment:
Doesn't this effectively force cache expiration on each get, so you don't really get the benefits of having a cache?

Reply:
no. if you read the documentation on cache tools the behavior is more or less what i described. expire only drops those entries that have timed out. all other entries will remain. clear is the operation to drop all cache entries no matter the timing.